docs/architecture/adr-060-baseapp++.md

```diff
@@ -0,0 +1,200 @@
+	// upon failure.
+	Insert(T) error
+	// Reap returns the next available transaction in the mempool, returning an error
+	// upon failure. The notion of 'available' or 'next' is defined by the application's
+	// mempool implementation.
+	Reap() (T, error)
+	// ReapMaxBytes performs the the same function as Reap, except it returns
+	// multiple transactions limited by their total size.
+	ReapMaxBytes(n int) ([]T, error)
+	// NumTxs returns the number of transactions currently in the mempool.
+	NumTxs() int
+}
+```
+
+We will define an implementation of `Mempool[T MempoolTx]` that will cover a
+majority of application use cases. Namely, it will prioritize transactions by
+priority and transaction sender, allowing for multiple prioritized transactions
+from the same sender. The mempool will be defined as a wrapper around a simple
+priority queue using a max binary heap, along with additional indexes/metadata
+to store senders and their nonces.
+
+Transaction reaping will essentially happen via a two-phase approach:
+
+1. Reap one or more transactions from the priority queue and collect them into 
+   one of two buffers -- _valid_ and _invalid_.
+2. For transactions that DO NOT violate the nonce validation, they are included
+   in the _valid_ buffer.
+3. For transactions that DO violate the nonce validation, they are included in
+   the _invalid_ buffer.
+4. Continue this process until the desired number of valid transactions are
+   reaped or until the mempool is empty.
+5. Provide Tendermint the list of all transactions from the _valid_ buffer.
+6. Re-insert all transactions, from both buffers, back into mempool. This is to
+   ensure we do not discard transactions from the mempool in case `ProcessProposal`
+   fails.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue   *PriorityQueue[MempoolTx]
+	senders map[string][]int64
+	// ...
+}
+```
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
+perform a request to the application, via `PrepareProposalRequest`, with a certain
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `PrepareProposalRequest`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `PrepareProposalRequest`, the application will
+ignore _all_ transactions sent to it in the request and instead reap up to
+`PrepareProposalRequest.max_tx_bytes` from it's own mempool. There is no need to
+execute the transactions for validity as they have already passed CheckTx.
+
+### `ProcessProposal`
```

Comment:
There is zero difference between `CheckTx` and `ProcessProposal` when it comes to executing txs, apart from the base state used. There isn't anything custom or special about `ProcessProposal` with regards to AnteHandlers. So I'm not sure I completely follow the statement?

Reply:
> Further, each transaction is run through the handlers sequentially and therefore the Application never has an opportunity to validate things like ordering of the transactions in the proposal.

Can you define order validation?

Reply:
For example, we'd like to enforce via `ProcessProposal` that validators have proposed certain transactions in a particular order in the block. These are transactions with a single bespoke message _only_ submitted by the proposer. That is – they are never broadcasted by normal clients or gossiped in the mempool.

To validate these types of transactions/messages it sounds like we would need to write custom Antehandlers for each, because some of them read from state and thus the state needs to be validated.

At the very least, I think we will need something like `ctx.IsProcessProposal()` to determine whether this transaction is being run for `ProcessProposal`.

I am not sure how we would enforce an ordering or mandatory inclusion of a particular transaction in the proposal, because there would be no way to determine if we are processing the last transaction in the proposal.

Reply:
I think I understand what you're getting at. If you're requesting that we define `IsProcessProposal bool` on the `Context` type so you can know what "mode" you're in during the AnteHandler execution, then yes, we can absolutely define that API 👍 

Is this what you mean/want?

Reply:
Awesome. Yes that would be great. However there is one more thing which I am not quite sure how to solve with this proposed implementation. It's the final point I made in my previous comment.

> I am not sure how we would enforce an ordering or mandatory inclusion of a particular transaction in the proposal, because there would be no way to determine if we are processing the last transaction in the proposal within the context of an Antehandler.

My comment may have been confusing so I can try to clarify further. We have some two requirements that I think are important to communicate.
1. We want to enforce the proposer always maintain a specific ordering of certain types of transaction in the block. For example, an Oracle Price update transaction must always be the last transaction in the block, else the validator should reject the proposal. Obviously, we will reorder these transactions how we expect them in `PrepareProposal` using our custom Mempool implementation which implements your interface. This ordering requirement seems possible for us to hack together using your proposed AnteHandler implementation. We may just have to maintain a reference somewhere to the previous transaction run through the AnteHandler.
2. We want to enforce the proposer always include some "mandatory" transactions, else the validator should reject the proposal. This is the one I am less sure how we would do.

Can you think of a way to write an AnteHandler that could verify that some "mandatory" transaction was included in the Proposal during `ProcessProposal` in your proposed implementation?

One way I can think to do it is:
1. Write a custom AnteHandler.
2. For each transaction run through the AnteHandler, check if `IsProcessProposal` is true. If this transaction is a "mandatory" transaction, flip some global flag (or set a value on the Context).
4. Upon processing the final `Msg` in the proposal, check the global flag. If no "mandatory" `Msg` has been seen, reject the proposal.

Specifically it's step 3 that I am unsure about, as I don't think (within the context of this AnteHandler) we would know if we could if we are seeing the last transaction in the proposal. Does this make sense?

Reply:
> 1. ... For example, an Oracle Price update transaction must always be the last transaction in the block, else the validator should reject the proposal. Obviously, we will reorder these transactions how we expect them in PrepareProposal using our custom Mempool implementation which implements your interface. This ordering requirement seems possible for us to hack together using your proposed AnteHandler implementation.

Tx ordering and tx validation are two entirely separate things. You have complete control over how to order txs as you've pointed out. I'm having trouble understanding where the AnteHandler comes into play here? Note, keep in mind that txs are already executed via CheckTx and the AnteHandler was already executed for them. This all happens prior to `PrepareProposal`.

> 2. We want to enforce the proposer always include some "mandatory" transactions, else the validator should reject the proposal. This is the one I am less sure how we would do.

Pretty sure you can do this in your mempool implementation?

> Can you think of a way to write an AnteHandler that could verify that some "mandatory" transaction was included in the Proposal during ProcessProposal in your proposed implementation? 

Your proposal seems fine I think. Keep in mind, you can write your AnteHandlers however you wish. In fact, when `ProcessProposal` is called, you already have the **entire** set of `TxRecords` and you can easily pass these to your AnteHandler or do whatever you wish with them.

Reply:
> Pretty sure you can do this in your mempool implementation?

Right. To be clear, this has nothing to do with `PrepareProposal`, only `ProcessProposal`. That is to say – we want to validate that the proposer is being honest (ordering transactions correctly, and including "mandatory" transactions) and reject the proposal if they are not.

> you already have the entire set of TxRecords and you can easily pass these to your AnteHandler or do whatever you wish with them.

This might be the piece I am missing. I am not sure where this set of `TxRecords` would live. How can would we reference them from within the `AnteHandler`. Would they be on the `Context`?

Reply:
You pass them to the AnteHandler call. `ProcessProposal` will execute the AnteHandler. You can set the records on the `Context`.

An additional approach we can consider is to modify the `AnteHandler` method signature to now also accept a series of `TxRecords`? It's not clear to me the best approach. I don't want to get too much into the weeds in this ADR with regards to the `AnteHandler`, but it might be that when it comes to the implementation we modify the `AnteHandler` method signature.

Reply:
Update: See this comment by @sergio-mena https://github.com/cosmos/cosmos-sdk/pull/12838#discussion_r944813538

This looks like it might be an issue for you :-/

Reply:
cc/ myself

Thanks @alexanderbez for noting the comment! FWIW, we understand the potential liveness implication associated with rejecting the proposal as noted in the linked comment as well as Requirement 6 & 7 in the [ABCI++ App Req doc](https://docs.tendermint.com/master/spec/abci++/abci++_app_requirements.html).

-------------

Going back to "how to validate the `TxRecords` ordering in `ProcessProposal`" topic, how I imagine things would work is:

1. `ProcessProposal` is called with `[]TxRecord`
2. for each `TxRecord` in `[]TxRecord`     
           (2a) call AnteHandler on a single `TxRecord`

Potentially, there are a few different ways to perform a "validate the ordering" check:

* Option A: between 1 and 2, call "validate the msg ordering" func on `[]TxRecord` and allow the application to define this func
* Option B: inside the loop in 2, AnteHandler knows which index the current `TxRecord` is and what msgs we've seen so far (i.e. all the msgs that were part of previous `TxRecord`s)
* Option C: instead of calling AnteHandler on a single `TxRecord`, allow AnteHandler to accept the entire list of `TxRecord` (i.e. `[]TxRecord`) and the AnteHandler can call "validate the msg ordering" on `[]TxRecord`

I think it would be nice to be able to do Option A, but not sure if this can be supported. Alternatively, Option B or Option C would be okay too (I believe Bryce was asking how B and C would actually work)

Let me know if I am misunderstanding anything!

Reply:
Your proposal requires no changes to the AnteHandler as it already takes a `tx` as an argument. Recall, a `TxRecord` is just a bytes blob which is decoded into a `Tx`.

> Option A: between 1 and 2, call "validate the msg ordering" func on []TxRecord and allow the application to define this func

I don't think this is viable as this would require the execution to happen in `ProcessProposal` which is custom to your app. It's not general enough.

I would recommend option B honestly.

Reply:
> I don't think this is viable as this would require the execution to happen in ProcessProposal which is custom to your app. It's not general enough.

Could you help elaborate on what you mean by "require the execution to happen"? In this specific example, I think the "validate" would iterate through the transactions and inspect msgs across all transactions. Does this count as an execution?

> I would recommend option B honestly.

With the current AnteHandler impl, is it possible to know, for example, "all the msgs that were part of previous transactions" inside AnteHandler? 

------------
Based on [this discussion](https://github.com/cosmos/cosmos-sdk/pull/12838#discussion_r948229670), it seems like `ProcessProposal` interface will be generic enough that the app has the ability to insert custom logic (something like "validate the ordering of all msgs/txns"). Is this a correct understanding? Does this mean Option A is  doable?

Reply:
The new design [here](https://github.com/cosmos/cosmos-sdk/pull/12838#discussion_r948229670) addresses all of the concerns from the dYdX side around `ProcessProposal`. I believe this thread can be resolved.

Reply:
This entire convo is now moot as we've now modified the ADR to allow the app to entirely define `ProcessProposal`. Resolving.